karrot/offers/api.py

```diff
@@ -0,0 +1,146 @@
+
+        JSON part:
+            {
+                "name": "foo",
+            }
+        "image" part:
+            <some binary content for the image>
+
+    OR, as you have to do in client JS:
+
+        const document = { name: "foo" }
+        const imageBlob = getImageBlobFromWhereever()
+        const data = new FormData()
+        data.append(
+          'document',
+          new Blob(
+            [JSON.stringify(document)],
+            { type: 'application/json' },
+          )
+        )
+        data.append('image', imageBlob, 'image.jpg')
+
+    """
+    def parse(self, stream, media_type=None, parser_context=None):
+        data = {}
+        parsed = MultiPartParser.parse(self, stream, media_type, parser_context)
+
+        # Find any JSON content first
+        for name, content in parsed.files.items():
+            if content.content_type != 'application/json':
+                continue
+            data.update(**json.load(content.file))
+
+        # Now get any other content
+        for name, content in parsed.files.items():
+            if content.content_type == 'application/json':
+                continue
+            # name is the path into the object to assign
+            glom.assign(data, name, content)
+
+        return data
+
+
+class IsOfferUser(BasePermission):
+    """Is the user the owner of the offer they wish to update?"""
+
+    message = _('You are not the owner of this offer')
+
+    def has_object_permission(self, request, view, offer):
+        if view.action != 'partial_update':
+            return True
+        return request.user == offer.user
+
+
+class OfferViewSet(
+        mixins.CreateModelMixin,
+        mixins.RetrieveModelMixin,
+        PartialUpdateModelMixin,
+        mixins.ListModelMixin,
+        GenericViewSet,
+        RetrieveConversationMixin,
+):
+    serializer_class = OfferSerializer
+    queryset = Offer.objects
+    filter_backends = (filters.DjangoFilterBackend, )
+    filterset_fields = (
+        'group',
+        'status',
+    )
+    pagination_class = OfferPagination
+    permission_classes = (
+        IsAuthenticated,
+        IsOfferUser,
+    )
+    parser_classes = [JSONWithFilesMultiPartParser, JSONParser]
+
+    def get_queryset(self):
+        return self.queryset.filter(
+            Q(group__members=self.request.user),
+            Q(user=self.request.user) | ~Q(status=OfferStatus.DISABLED.value),
+        ).distinct()
+
+    @action(
+        detail=True,
+    )
+    def conversation(self, request, pk=None):
+        """Get conversation ID of this offer"""
+        return self.retrieve_conversation(request, pk)
+
+    @action(
+        detail=True,
+        methods=['GET'],
+        permission_classes=(),
+    )
+    def image(self, request, pk=None):
+        image = OfferImage.objects.filter(offer=pk).first()
+        if not image:
+            raise Http404()
+        return HttpResponseRedirect(redirect_to=image.image.url)
```

Comment:
vue-croppa is resizing images before upload, I just chose a 10mb image for my dev.karrot.world profile, and the "fullsize" image ends up as https://dev.karrot.world/media/user__photos/photo_cntAK1q.jpeg (~212kb), and the request meta in firefox confirms that upload content length is 217423.

Reply:
... technically they could upload it outside of karrot UI and get a bigger photo I guess ...

Reply:
vue-croppa usually works fine, but I sometimes encountered really big images around 1MB (I think it was PNG files instead of JPEG). And the API allows all image formats and sizes anyways. But as this problem already exists with the other image usages (group photo and user profile picture), you don't have to solve it for the offers. Just to keep in mind that it _could_ become problematic to default to the fullsize version.